### ISSUE TITLE
Connection to MSSQL Server

### ISSUE THREAD

**Issue by movinghands** [gunthercox/ChatterBot#1144]
Hi,

May I know the syntax for setting an MSSQL server as the database storage adapter? 

So far I've tried these 3 methods of setting the database URI in the Chatterbot configuration:

**1)**

    chatbot = ChatBot(
        'bot',
        storage_adapter='chatterbot.storage.SQLStorageAdapter',
        database_uri='mssql+pyodbc://username:password@dsn'
    )

**2)**

    chatbot = ChatBot(
        'bot',
        storage_adapter='chatterbot.storage.SQLStorageAdapter',
        database_uri='DRIVER={ODBC Driver 11 for SQL Server};Server=server;Database=database;UID=uid;PWD=pwd;'
    )

**3)**

    chatbot = ChatBot(
        'bot',
        storage_adapter='chatterbot.storage.SQLStorageAdapter',
        database_uri=urllib.parse.quote_plus('DRIVER={ODBC Driver 11 for SQL Server};Server=server;Database=database;UID=uid;PWD=pwd;')
    )

None of these methods were successful. Please advice on how I should solve this, thank you!

**Comment by vkosuri** [comment]
@movinghands I think there is no support Microsoft SQL server, current supports are Django, Sqlalchemy, Mongo DB

I you really required to support MSSQL then you have to write your own storage adapter. You could beautiful documentation here http://chatterbot.readthedocs.io/en/stable/storage/create-a-storage-adapter.html

**Comment by movinghands** [comment]
I see. SQLalchemy does have MSSQL server support though, which was what I tried. But I'm not sure if I've got the syntax correct for the connection URI. Thank you for replying!

**Comment by vkosuri** [comment]
Thank you could try like this as described here http://docs.sqlalchemy.org/en/latest/faq/connections.html#i-m-on-sql-server-how-do-i-turn-those-rollbacks-into-commits

```
engine = create_engine('mssql://scott:tiger@mydsn', pool=QueuePool(reset_on_return='commit'))
```

**Comment by movinghands** [comment]
@vkosuri Where do I place the engine?

**Comment by vkosuri** [comment]
I didn't see that, the method you already tried, If not can you try?

``` Python
chatbot = ChatBot(
    'bot',
    storage_adapter='chatterbot.storage.SQLStorageAdapter',
    database_uri='mssql://scott:tiger@mydsn'
)
```

**Comment by movinghands** [comment]
@vkosuri This is the error message I get:

> AttributeError: 'MSDialect_pyodbc' object has no attribute 'default_schema_name'

**Comment by vkosuri** [comment]
I am nit sure about this error, I googled a bit, found this one https://groups.google.com/forum/m/#!topic/sqlalchemy/jK1TcAbcyxE

**Comment by movinghands** [comment]
@vkosuri SQLalchemy does have support for pyodbc and MSSQL, as seen [here](http://docs.sqlalchemy.org/en/latest/dialects/mssql.html#module-sqlalchemy.dialects.mssql.pyodbc). But the issue seems to be on Chatterbot's side instead — I am able to connect to my database using SQLalchemy

The error occurs when the database URI is passed into Chatterbot's storage adapter database URI parameter

**Comment by vkosuri** [comment]
Thanks, just curiosity, how your connecting sqlalchemy using mssql. Not from chatterbot

**Comment by movinghands** [comment]
@vkosuri Using the engine, like so:

> engine = create_engine('mssql+pyodbc://username:password@dsn')

However, to my understanding, I cannot use the engine when connecting with Chatterbot

**Comment by vkosuri** [comment]
Yeah I see, chatterbot adding sqlite prefix, may be you could remove and try with this ``engine = create_engine('mssql+pyodbc://username:password@dsn')`` 

https://github.com/gunthercox/ChatterBot/blob/6790455163a14f6ddb084bfffda6e64e2b422f84/chatterbot/storage/sql_storage.py#L58

https://github.com/gunthercox/ChatterBot/blob/6790455163a14f6ddb084bfffda6e64e2b422f84/chatterbot/storage/sql_storage.py#L60

**Comment by movinghands** [comment]
@vkosuri That seems optimistic; I'll give it a try on Monday. Thank you for your reply :)

**Comment by vkosuri** [comment]
My assumption are wrong, it won't goes into ``database_name`` condition part, because you are passing ``database_uri``, Give a try and let me know?

**Comment by movinghands** [comment]
@vkosuri Yes, that seems to be the case unfortunately. My URI is properly passed to the engine

**Comment by movinghands** [comment]
This issue is still a problem. Any help would be greatly appreciated!
